### PR TITLE
feat!: Traverse the Noosphere vast

### LINKED ISSUE #2: Change `Bundle` to enable a streaming interface
We have a `Bundle` construct that is used to gather up the sparse blocks needed to represent a given set of revisions. It is composable / mergeable with itself, which makes it nice for slurping up incremental sets of blocks for synchronization between client and gateway.

Unfortunately, gathering blocks is currently a blocking (in the order of operations) routine, and must be completed before a `Bundle` can be serialized or deserialized.

This means there may be cases where we allocate a lot of memory just to move a bunch of tiny blocks from one storage location to the next. It also means we may defer serialization for a much longer time than necessary while we collect blocks from storage. Let's fix this by changing the `Bundle` to use a streaming interface. Maybe it can be rebuilt around [`AsyncRead`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncRead.html)/[`AsyncWrite`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWrite.html)?

### LINKED ISSUE #269: Enable traversal from one sphere to the next
When we store sphere data for spheres that belong to other authors besides the configured author - where "author" is essentially a key holder - we store it in the same storage as the author's own sphere. So, for a given author accessing a third-party sphere to reliably access the blocks that have been previously fetched, we need a handle to their credentials as well as the backing storage layer.

Sphere "traversal" is the process by which the author of one sphere accesses another sphere via the context of their own sphere.

In order to implement this feature, we need to perform the following (roughly):

 1. Initialize the origin `SphereContext` for the sphere the author is starting from
 2. Begin traversal by specifying the DID of the destination sphere
 3. If a gateway is configured, create a pending request to the gateway to replicate the destination sphere locally
 4. If the destination sphere data is available locally, immediately resolve it; otherwise, block on the pending request to the gateway before resolving it
 5. Create a `SphereContext` for the destination sphere configured with the same author and storage backend as the origin `SphereContext`

### PR DESCRIPTION
**Pull request by cdata**
Building on the work in #278 and #283 , this change implements our first take on sphere traversal as outlined in #269. By invoking `SphereContext::traverse_by_petname`, a user can move from one sphere to the next with minimum information about the sphere's data format or its relative location on the network.

This change implements sphere traversal in Rust, but stops short of adding an FFI; I'll add the FFI in a smaller follow-up change.

Important note: this change breaks our HAMT's data layout. In the past, HAMT values were stored and encoded directly inside a `Vec<KeyValuePair>`. As of this change, `KeyValuePair` internally holds a link to the data, not the data itself. This reduces the amount of data we have to store, but increases the number of lookups we have to do from storage in order to read the values of a HAMT. We are compensated for this trade-off by transferring less over the wire during replication (and it also simplifies the logic of determining what to stream when replicating).

Fixes #269 
Fixes #2

### CODE REVIEW

File: rust/noosphere-gateway/src/route/replicate.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    })?;
+
+    let db = sphere_context
+        .sphere_context()
+        .await
+        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?
+        .db()
+        .clone();
+    let store = BlockStoreRetry::new(
+        IpfsStore::new(db, Some(ipfs_client)),
```
Comment (jsantell) [inline rust/noosphere-gateway/src/route/replicate.rs:60]: `ipfs_client` isn't an Arc or Mutex, how does that work across many handlers? (will need to integrate an ipfs_client in the NS worker as well)

File: rust/noosphere-gateway/src/worker/syndication.rs [hunk 2]
```diff
@@ -89,15 +91,19 @@ where
         };
         let checkpoint_key = format!("syndication/kubo/{kubo_identity}");
 
+        debug!("IPFS node identified as {}", kubo_identity);
```
Comment (jsantell) [inline rust/noosphere-gateway/src/worker/syndication.rs:94]: We should set some RUST_LOG level for gateways in the cluster to verify and get insight in deployment

File: rust/noosphere-gateway/src/worker/syndication.rs [hunk 3]
```diff
@@ -89,15 +91,19 @@ where
...
+
         // Take a lock on the `SphereContext` and look up the most recent
         // syndication checkpoint for this Kubo node
         let (sphere_revision, ancestor_revision, mut syndicated_blocks, db) = {
             let db = {
                 let context = context.sphere_context().await?;
                 context.db().clone()
             };
+            info!("CHECKPOINT A");
```
Comment (jsantell) [inline rust/noosphere-gateway/src/worker/syndication.rs:103]: nit: low-context logs

**Review by jsantell — APPROVED** [review]
Nice work! Can't wait to see this in action

File: rust/noosphere-gateway/src/worker/mod.rs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+mod name_system;
+mod replication;
+mod syndication;
+
+pub use name_system::*;
+pub use replication::*;
```
[the remark this answers is not in the document]
Comment (cdata) [reply rust/noosphere-gateway/src/worker/mod.rs:6]: Will remove

File: rust/noosphere-gateway/src/route/replicate.rs (hunk 1 above)
Comment (cdata) [reply rust/noosphere-gateway/src/route/replicate.rs:60]: It is `Clone`; it gets cloned for every handler invocation.

File: rust/noosphere-gateway/src/worker/syndication.rs (hunk 2 above)
Comment (cdata) [reply rust/noosphere-gateway/src/worker/syndication.rs:94]: SG. By default it is set for `orb`, but the scope is somewhat limited in scope to a few crates.

File: rust/noosphere-storage/src/tap.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
+use crate::BlockStore;
+use anyhow::Result;
+use async_trait::async_trait;
+use cid::Cid;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+
+#[derive(Clone)]
+pub struct BlockStoreTap<S>
```
[the remark this answers is not in the document]
Comment (cdata) [reply rust/noosphere-storage/src/tap.rs:8]: Sorry, I wrote doc comments but I somehow managed not to save in my editor. Look for them shortly!

File: rust/noosphere-gateway/src/route/replicate.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+/// Invoke to get a streamed CARv1 response that represents all the blocks
+/// needed to manifest the content associated with the given CID path parameter.
+/// The CID should refer to the memo that wraps the content. The content-type
+/// header is used to determine how to gather the associated blocks to be
+/// streamed by to the requesting client. Invoker must have authorization to
+/// fetch from the gateway.
+pub async fn replicate_route<C, K, S>(
+    authority: GatewayAuthority<K>,
+    // NOTE: Cannot go from string to CID via serde
```
[the remark this answers is not in the document]
Comment (cdata) [reply rust/noosphere-gateway/src/route/replicate.rs:27]: I think we would have to use a newtype in this case, unless serde field attributes work on function args... but I don't think they do.

File: rust/noosphere/src/noosphere.rs [hunk 7]
```diff
@@ -113,11 +113,9 @@ impl NoosphereContext {
         }
     }
 
-    fn ipfs_gateway_url(&self) -> Option<&Url> {
+    fn ipfs_api(&self) -> Option<&Url> {
```
[the remark this answers is not in the document]
Comment (cdata) [reply rust/noosphere/src/noosphere.rs:116]: Reverted

File: rust/noosphere-sphere/src/replication.rs [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+                revocations_result??;
+            }
+            Some(_) => {
+                let stream = BodyChunkDecoder(&memo.body, &store).stream();
+
+                drop(store);
+
+                tokio::pin!(stream);
+
+                'decode: while let Some(_) = stream.try_next().await? {
```
[the remark this answers is not in the document]
Comment (cdata) [reply rust/noosphere-sphere/src/replication.rs:92]: :sparkles: :label: :sparkles:

**Review by jsantell — APPROVED** [review]

**Pull request merged**